OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs

```diff
@@ -1085,6 +1085,40 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
 					}
 				}
 
+				if (engineVersion < 20150607)
+				{
+					// Add WithRankDecoration to all actors using GainsExperience
+					var ge = node.Value.Nodes.FirstOrDefault(n => n.Key == "GainsExperience");
+					if (ge != null)
+					{
+						var nodeUpgrades = ge.Value.Nodes.FirstOrDefault(n => n.Key == "Upgrades");
+						var upgrades = nodeUpgrades != null ? nodeUpgrades.Value.Nodes.Count() : 4;
+
+						if (upgrades != 0)
+						{
+							foreach (var nodeUpgrade in nodeUpgrades.Value.Nodes)
+								nodeUpgrade.Value.Value = "rank" + (string.IsNullOrEmpty(nodeUpgrade.Value.Value) ? null : ", ") + nodeUpgrade.Value.Value;
+
+							var nodeChPal = ge.Value.Nodes.FirstOrDefault(n => n.Key == "ChevronPalette");
+							var chPal = nodeChPal != null && !string.IsNullOrEmpty(nodeChPal.Value.Value) ? nodeChPal.Value.Value : "effect";
+							ge.Value.Nodes.Remove(nodeChPal);
```

Comment:
`upgrades == 0` is only possible if the `Upgrades` dictionary is empty. This renders ´GainsExperience` useless and, as far as I can understand, should be equivalent to simply using `-GainsExperience`. I've seen this in `mods/ra/maps/desert-shellmap/map.yaml` and I'm not sure why it is done this way...

So if I get this right, in this case not only that the palette is no longer needed, as you say, but the whole node can be replaced by `-GainsExperience`...